api/src/main/java/org/apache/iceberg/metrics/ScanReport.java

```diff
@@ -102,87 +113,364 @@ public Builder withSnapshotId(long newSnapshotId) {
-      Preconditions.checkArgument(null != filter, "Invalid expression filter: null");
       Preconditions.checkArgument(null != projection, "Invalid schema projection: null");
       Preconditions.checkArgument(null != scanMetrics, "Invalid scan metrics: null");
-      return new ScanReport(tableName, snapshotId, filter, projection, scanMetrics);
+      return new ScanReport(tableName, snapshotId, projection, scanMetrics);
+    }
+  }
+
+  /** A serializable version of a {@link Timer} that carries its result. */
+  public static class TimerResult {
+    private final String name;
+    private final TimeUnit timeUnit;
+    private final Duration totalDuration;
+    private final long count;
+
+    static TimerResult fromTimer(Timer timer) {
+      Preconditions.checkArgument(null != timer, "Invalid timer: null");
+      if (Timer.NOOP.equals(timer)) {
+        return null;
+      }
+      return new TimerResult(timer.name(), timer.unit(), timer.totalDuration(), timer.count());
+    }
+
+    TimerResult(String name, TimeUnit timeUnit, Duration totalDuration, long count) {
+      Preconditions.checkArgument(null != name, "Invalid timer name: null");
+      Preconditions.checkArgument(null != timeUnit, "Invalid time unit: null");
+      Preconditions.checkArgument(null != totalDuration, "Invalid duration: null");
+      this.name = name;
+      this.timeUnit = timeUnit;
+      this.totalDuration = totalDuration;
+      this.count = count;
+    }
+
+    public String name() {
+      return name;
+    }
+
+    public TimeUnit timeUnit() {
+      return timeUnit;
+    }
+
+    public Duration totalDuration() {
+      return totalDuration;
+    }
+
+    public long count() {
+      return count;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(name())
+          .add("duration", totalDuration())
+          .add("count", count)
+          .add("timeUnit", timeUnit)
+          .toString();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+
+      TimerResult that = (TimerResult) o;
+      return count == that.count
+          && Objects.equal(name, that.name)
+          && timeUnit == that.timeUnit
+          && Objects.equal(totalDuration, that.totalDuration);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(name, timeUnit, totalDuration, count);
+    }
+  }
+
+  /** A serializable version of a {@link Counter} that carries its result. */
+  public static class CounterResult {
+    private final String name;
+    private final MetricsContext.Unit unit;
+    private final long value;
+
+    static CounterResult fromCounter(Counter<Long> counter) {
+      Preconditions.checkArgument(null != counter, "Invalid counter: null");
+      if (LongCounter.NOOP.equals(counter)) {
```

Comment:
Do we need to import this `LongCounter`? build is failing locally because of this.

Reply:
Sorry for the troubles, this was unfortunately caused by a silent merge conflict. Build was fixed by https://github.com/apache/iceberg/pull/5506 and just merged a few minutes ago.